### PR TITLE
Hoist loop-invariant leading lets out of canonical Matchless recursion loops

### LINKED ISSUE #1738: Hoist loop-invariant lets out of Loop bodies
## Summary
Follow-up optimization from #1728: hoist loop-invariant bindings that are computed inside a `Loop` body to outside the loop when safe.

## Motivation
Current loop normalization can lift invariant **loop arguments**, but does not hoist invariant `let` bindings that appear in the loop body.

Example pattern:

```bosatsu
loop(x = a, y = b):
  z = f(100)
  if x > 0:
    recur(x - 1, y + z)
  else:
    (0, y)
```

`z = f(100)` is loop-invariant and could be hoisted above the loop.

## Scope
- detect loop-body lets that are invariant across all recur paths
- hoist eligible bindings above the `Loop` while preserving semantics
- add normalization tests for positive and negative cases (effects/escaping refs/shadowing)

Related: #1728

### PR DESCRIPTION
**Pull request by johnynek**
Implemented issue #1738 in Matchless and tests.

Code changes:
- Added `hoistInvariantLoopLets` in `core/src/main/scala/dev/bosatsu/Matchless.scala`.
- Updated `Matchless.fromLet` to run `hoistInvariantLoopLets` before `reuseConstructors`.
- The new pass is intentionally conservative and only rewrites canonical recursion-loop scaffolding:
  `Always(SetMut(condMut, TrueExpr), WhileExpr(isTrueExpr(condMut), effectExpr, resultMut))`.
- Inside eligible loop bodies, it inspects only the contiguous leading immutable `Let` prefix and hoists bindings only when all of the following hold:
  - binding is used in residual loop body
  - value is pure (`!hasSideEffect`)
  - value has no mutable reads (`LocalAnonMut` barrier)
  - value contains no nested `WhileExpr`
  - value passes profitability (`!triviallyCheap` and `exprWeight >= 4`)
  - defensive name-shadowing guard for `Right(name)` binders against loop condition references
- Hoisted bindings preserve original ordering so dependency chains remain correct.

Test changes (`core/src/test/scala/dev/bosatsu/MatchlessTests.scala`):
- Added direct pass tests for:
  - positive hoist of a top-level invariant loop-body let
  - positive hoist of dependent leading lets in order
  - negative: cheap alias / mutable read / side-effectful value
  - negative: lambda value capturing mutable state
  - negative: let inside conditional branch
  - negative: non-canonical loop condition not rewritten
- Added integration test proving `Matchless.fromLet` hoists invariant let computation out of `WhileExpr.effectExpr`.

Validation:
- Re-ran required checks from `coding_style.md` on final code:
  - `sbt "coreJVM/test:compile"`
  - `sbt "coreJVM/test; cli/test"`
- Both passed.

Fixes #1738

Implements design doc: [docs/design/1738-hoist-loop-invariant-lets-out-of-loop-bodies.md](https://github.com/johnynek/bosatsu/blob/main/docs/design/1738-hoist-loop-invariant-lets-out-of-loop-bodies.md)

Design source PR: https://github.com/johnynek/bosatsu/pull/1785

### CODE REVIEW

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 1]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
     StackSafe.onStackOverflow(recurExpr(expr, init)._1)(expr)
   }
 
+  private val HoistInvariantLoopLetsMinWeight: Int = 4
+
+  // Hoist loop-invariant leading lets out of canonical recursion loops.
+  private[bosatsu] def hoistInvariantLoopLets[A](expr: Expr[A]): Expr[A] = {
+    def letBinds(
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: move this to a method on `object Let` and double check for duplicates. I think we have written this function before.

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 2]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+            loopBool(in)
+          case TrueConst =>
+            false
+        }
+
+      loopExpr(ex)
+    }
+
+    def referencesBindable(expr: Expr[A], target: Bindable): Boolean = {
+      def loopExpr(e: Expr[A], isShadowed: Boolean): Boolean =
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: if `isShadowed == true` we can return false right? We can never return true when that is true right? Can't we stop the recursion at that point?

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 3]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+    }
+
+    def referencesBindableInBool(
+        cond: BoolExpr[A],
+        target: Bindable
+    ): Boolean =
+      referencesBindable(Always(cond, UnitExpr), target)
+
+    def referencesLocalAnon(expr: Expr[A], target: Long): Boolean = {
+      def loopExpr(e: Expr[A], isShadowed: Boolean): Boolean =
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: same comment about shadowing, can't we check `if (isShadowed)` and return false?

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 4]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+          case LetMutBool(_, in) =>
+            1 + loopBool(in)
+          case TrueConst =>
+            1
+        }
+
+      loopExpr(expr)
+    }
+
+    def containsWhileExpr(ex: Expr[A]): Boolean = {
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: this can be on `object Expr` and be tested directly.

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 5]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+          case LetMutBool(_, in) =>
+            loopBool(in)
+          case TrueConst =>
+            false
+        }
+
+      loopExpr(ex)
+    }
+
+    def readsMutableExpr(ex: Expr[A]): Boolean = {
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: this could be on `object Expr` so we can test directly and maybe share on different optimizations.

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 6]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+          case LetMutBool(_, in) =>
+            loopBool(in)
+          case TrueConst =>
+            false
+        }
+
+      loopExpr(ex)
+    }
+
+    def referencesBindable(expr: Expr[A], target: Bindable): Boolean = {
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: this should be on `object Expr` for direct testing and sharing across optimizations.

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 7]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+          case LetMutBool(_, in) =>
+            loopBool(in, isShadowed)
+          case TrueConst =>
+            false
+        }
+
+      loopExpr(expr, isShadowed = false)
+    }
+
+    def referencesBindableInBool(
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: move to `object BoolExpl`

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 8]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+        cond: BoolExpr[A],
+        target: Bindable
+    ): Boolean =
+      referencesBindable(Always(cond, UnitExpr), target)
+
+    def referencesLocalAnon(expr: Expr[A], target: Long): Boolean = {
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: move to `object Expr`

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 9]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+          case LetMutBool(_, in) =>
+            loopBool(in, isShadowed)
+          case TrueConst =>
+            false
+        }
+
+      loopExpr(expr, isShadowed = false)
+    }
+
+    def usesBinding(expr: Expr[A], arg: Either[LocalAnon, Bindable]): Boolean =
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: move to `object Expr`

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 10]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+          referencesLocalAnon(expr, id)
+      }
+
+    def isTriviallyCheap(value: Expr[A]): Boolean =
+      value match {
+        case _: CheapExpr[?] =>
+          true
+        case cons: ConsExpr =>
+          cons.arity == 0
+        case App(fn: ConsExpr, args) =>
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala]: this is an allocation. Why is this trivially cheap? If we could reduce an allocation it seems worth it. Maybe if the `args.length == 1` it's cheap, but otherwise I don't follow. Add a comment if you still think this makes sense.

File: core/src/main/scala/dev/bosatsu/Matchless.scala [hunk 11]
```diff
@@ -1085,6 +1085,466 @@ object Matchless {
...
+            true
+        }
+
+      noShadowingHazard &&
+      usesBinding(in, arg) &&
+      !hasSideEffect(value) &&
+      !readsMutableExpr(value) &&
+      !containsWhileExpr(value) &&
+      !isTriviallyCheap(value) &&
+      (exprWeight(value) >= HoistInvariantLoopLetsMinWeight)
```
Comment (johnynek) [inline core/src/main/scala/dev/bosatsu/Matchless.scala:1433]: add a comment here. Why do we not want to lift something cheap. Isn't everything a win?

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 1 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Moved the let-chain builder to `object Let` as `bindAll`, and de-duplicated the related helpers by using `Let.bindNamed` / `Let.bindAnons` where we previously repeated local `foldRight` let builders.

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 2 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Yes, agreed. `Expr.referencesBindable` now short-circuits with `if (isShadowed) false`, so we stop recursing as soon as the target is shadowed.

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 3 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Applied the same shadowing short-circuit to `Expr.referencesLocalAnon` (both expr and bool walkers now return `false` immediately when shadowed).

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 4 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Moved this to `Matchless.Expr.containsWhileExpr` and added direct test coverage in `MatchlessTests` (`Matchless.Expr helper predicates are directly testable`).

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 5 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Moved this to `Matchless.Expr.readsMutable` and added direct tests for it in `MatchlessTests`.

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 6 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Moved this to `Matchless.Expr.referencesBindable` (with shadow-aware shared helper form) so it is directly testable and reusable.

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 7 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Moved to `Matchless.BoolExpr.referencesBindable` and updated the hoist pass to use that bool-level helper directly.

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 8 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Moved to `Matchless.Expr.referencesLocalAnon` and added direct shadowing tests for it.

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 9 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: Moved to `Matchless.Expr.usesBinding`, updated hoist logic to call it, and added direct tests for both bindable and local-anon bindings.

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 10 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala]: I changed the behavior here: constructor applications are no longer treated as trivially cheap. `isTriviallyCheap` now only classifies `CheapExpr` and zero-arity constructors as cheap, and I added `hoists non-trivial constructor allocations` to lock this in.

File: core/src/main/scala/dev/bosatsu/Matchless.scala (hunk 11 above)
Comment (johnynek) [reply core/src/main/scala/dev/bosatsu/Matchless.scala:1433]: Added a comment at the profitability gate explaining why cheap bindings are not always hoisted: hoisting can increase live ranges and residual loop body size, so we require non-trivial work before moving.

**Comment by johnynek** [comment]
Implemented all requested refactors and comments on the PR thread with code changes plus direct tests for the new helper locations. I also updated the C codegen golden hash in `ClangGenTest` because this optimization changed generated output shape (`Ackermann` hash now `32eeba83ae74c2c8fa75703823fe7298`).

Checks run locally:
- `sbt "coreJVM/test:compile"`
- `sbt "coreJVM/testOnly dev.bosatsu.MatchlessTest"`
- `sbt "cli/testOnly dev.bosatsu.codegen.clang.ClangGenTest"`
- `sbt "coreJVM/test; cli/test"`
